### ISSUE TITLE
Memory spike

### ISSUE THREAD

**Issue by bowenwang1996** [near/nearcore#3725]
We observed several times that for nodes on mainnet that they sometimes experience a very drastic memory spike that causes the node to run out of memory in 10-20 minutes. @pmnoxx could you run a node on mainnet with your memory profiling tools to see if there is something abnormal?

- [x] write code for detecting memory spikes
- [ ] start dedicated node running on mainnet for week or two to see if there are any spikes
- [ ] wait for results

**Comment by pmnoxx** [comment]
I'll start the node today.

**Comment by pmnoxx** [comment]
I've been running a few nodes during last few days looks for any memory spikes. So far I found one such spike:
```
Dec 20 11:02:40.262  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 2505ms    
Dec 20 11:02:40.751  WARN near_memory_tracker::allocator: Thread 27632 reached new record of memory usage 3205MiB

 2: alloc::raw_vec::RawVec<T,A>::with_capacity_in
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:163:9
      alloc::raw_vec::RawVec<T>::with_capacity
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:93:9
      alloc::vec::Vec<T>::with_capacity
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:363:20
      alloc::slice::hack::to_vec
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/slice.rs:161:23
      alloc::slice::<impl [T]>::to_vec
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/slice.rs:397:9
      near_store::trie::state_parts::<impl near_store::trie::Trie>::combine_state_parts::{{closure}}
             at $HOME/repos/nearcore.measure-mem/core/store/src/trie/state_parts.rs:278:25
      core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:280:13
   3: core::option::Option<T>::map
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:450:29
      <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/mod.rs:924:9
      alloc::vec::Vec<T>::extend_desugared
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2433:35
      <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::spec_extend
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2356:9
      <alloc::vec::Vec<T> as alloc::vec::SpecFromIterNested<T,I>>::from_iter
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2139:9
      <alloc::vec::Vec<T> as alloc::vec::SpecFromIter<T,I>>::from_iter
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2162:9
   4: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2002:9
      core::iter::traits::iterator::Iterator::collect
             at $HOME/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1670:9
      near_store::trie::state_parts::<impl near_store::trie::Trie>::combine_state_parts
             at $HOME/repos/nearcore.measure-mem/core/store/src/trie/state_parts.rs:274:21
   5: <neard::runtime::NightshadeRuntime as near_chain::types::RuntimeAdapter>::confirm_state
             at $HOME/repos/nearcore.measure-mem/neard/src/runtime.rs:1313:28
   6: near_chain::chain::Chain::set_state_finalize
             at $HOME/repos/nearcore.measure-mem/chain/chain/src/chain.rs:1737:9
   7: near_client::sync::StateSync::sync_shards_status
             at $HOME/repos/nearcore.measure-mem/chain/client/src/sync.rs:637:27
   8: near_client::sync::StateSync::run
             at $HOME/repos/nearcore.measure-mem/chain/client/src/sync.rs:945:46
   9: near_client::client_actor::ClientActor::sync
             at $HOME/repos/nearcore.measure-mem/chain/client/src/client_actor.rs:1306:44
  10: near_client::client_actor::ClientActor::sync::{{closure}}
             at $HOME/repos/nearcore.measure-mem/chain/client/src/client_actor.rs:1386:17
      near_performance_metrics::actix::run_later::{{closure}}
             at $HOME/repos/nearcore.measure-mem/utils/near-performance-metrics/src/actix.rs:51:13
      <F as actix::utils::TimerFuncBox<A>>::call
             at $HOME/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.9.0/src/utils.rs:114:9
```

I modified near_memory_tracker to track the amount of memory allocated by each thread, I added code which print stack strace every time the amount of memory used by given thread increases by 512mb. I have seen a few similar stack traces printed, just like the one above.

It looks like ClientActor::sync is responsible for this memory spike.

**Comment by pmnoxx** [comment]
Here are some logs, which happened before this memory spike
```
Dec 20 11:01:20.441  INFO telemetry: Telemetry data could not be sent due to: Timeout out while waiting for response
Dec 20 11:01:25.345  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3135ms    
Dec 20 11:01:32.434  INFO stats: State [0: parts]  32/26/40 peers ⬇ 15.2MiB/s ⬆ 64.6kiB/s 0.00 bps 0 gas/s CPU: 373%, Mem: 2.8 GiB    
Dec 20 11:01:32.442  WARN sync: State sync didn't download the state for shard 0 in 10 seconds, sending StateRequest again    
Dec 20 11:01:41.625  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3020ms    
Dec 20 11:01:43.325  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 550ms peak_memory: 3MiB    
Dec 20 11:01:45.068  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 526ms peak_memory: 2MiB    
Dec 20 11:01:45.950  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 700ms peak_memory: 3MiB    
Dec 20 11:01:46.526  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 576ms peak_memory: 3MiB    
Dec 20 11:01:47.947  INFO stats: State [0: parts]  32/21/40 peers ⬇ 20.3MiB/s ⬆ 63.4kiB/s 0.00 bps 0 gas/s CPU: 346%, Mem: 2.9 GiB    
Dec 20 11:01:48.002  WARN sync: State sync didn't download the state for shard 0 in 10 seconds, sending StateRequest again    
Dec 20 11:01:48.482  INFO near_performance_metrics::actix: Slow function call run_later:1 chain/client/src/client_actor.rs:1383 took: 534ms    
Dec 20 11:01:51.105  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 868ms peak_memory: 4MiB    
Dec 20 11:01:52.273  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 1166ms peak_memory: 3MiB    
Dec 20 11:01:52.836  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 548ms peak_memory: 3MiB    
Dec 20 11:01:53.863  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 530ms peak_memory: 3MiB    
Dec 20 11:01:55.521  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3595ms    
Dec 20 11:01:56.243  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 510ms peak_memory: 1MiB    
Dec 20 11:01:58.143  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 514ms peak_memory: 3MiB    
Dec 20 11:02:07.887  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 539ms peak_memory: 4MiB    
Dec 20 11:02:10.534  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 550ms peak_memory: 2MiB    
Dec 20 11:02:11.026  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3304ms    
Dec 20 11:02:12.052  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 500ms peak_memory: 1MiB    
Dec 20 11:02:12.588  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 507ms peak_memory: 3MiB    
Dec 20 11:02:13.357  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 718ms peak_memory: 2MiB    
Dec 20 11:02:14.287  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 543ms peak_memory: 3MiB    
Dec 20 11:02:14.924  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 636ms peak_memory: 2MiB    
Dec 20 11:02:15.578  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 637ms peak_memory: 3MiB    
Dec 20 11:02:16.730  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 500ms peak_memory: 1MiB    
Dec 20 11:02:18.191  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 508ms peak_memory: 3MiB    
Dec 20 11:02:18.821  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 579ms peak_memory: 3MiB    
Dec 20 11:02:19.239  INFO near_performance_metrics::stats: Performance stats 38 threads    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:     Thread:1 ratio: 0.9088 {"near_client::sync::StateSync", "near_client::client_actor::ClientActor", "run_later", "near_telemetry::TelemetryActor", "telemetry"}:0     
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:737 cnt: 30 total: 11ms max: 0ms    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:1182 cnt: 25 total: 27ms max: 18ms    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:1383 cnt: 123 total: 1681ms max: 534ms    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:1396 cnt: 2 total: 40ms max: 30ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 chain/client/src/sync.rs:895 cnt: 3593 total: 509ms max: 34ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 chain/telemetry/src/lib.rs:70 cnt: 14 total: 15ms max: 9ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 near_network::types::NetworkClientMessages:0 cnt: 3460 total: 52242ms max: 1166ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 near_telemetry::TelemetryEvent:0 cnt: 2 total: 1ms max: 0ms    
Dec 20 11:01:20.441  INFO telemetry: Telemetry data could not be sent due to: Timeout out while waiting for response
Dec 20 11:01:25.345  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3135ms    
Dec 20 11:01:32.434  INFO stats: State [0: parts]  32/26/40 peers ⬇ 15.2MiB/s ⬆ 64.6kiB/s 0.00 bps 0 gas/s CPU: 373%, Mem: 2.8 GiB    
Dec 20 11:01:32.442  WARN sync: State sync didn't download the state for shard 0 in 10 seconds, sending StateRequest again    
Dec 20 11:01:41.625  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3020ms    
Dec 20 11:01:43.325  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 550ms peak_memory: 3MiB    
Dec 20 11:01:45.068  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 526ms peak_memory: 2MiB    
Dec 20 11:01:45.950  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 700ms peak_memory: 3MiB    
Dec 20 11:01:46.526  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 576ms peak_memory: 3MiB    
Dec 20 11:01:47.947  INFO stats: State [0: parts]  32/21/40 peers ⬇ 20.3MiB/s ⬆ 63.4kiB/s 0.00 bps 0 gas/s CPU: 346%, Mem: 2.9 GiB    
Dec 20 11:01:48.002  WARN sync: State sync didn't download the state for shard 0 in 10 seconds, sending StateRequest again    
Dec 20 11:01:48.482  INFO near_performance_metrics::actix: Slow function call run_later:1 chain/client/src/client_actor.rs:1383 took: 534ms    
Dec 20 11:01:51.105  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 868ms peak_memory: 4MiB    
Dec 20 11:01:52.273  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 1166ms peak_memory: 3MiB    
Dec 20 11:01:52.836  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 548ms peak_memory: 3MiB    
Dec 20 11:01:53.863  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 530ms peak_memory: 3MiB    
Dec 20 11:01:55.521  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3595ms    
Dec 20 11:01:56.243  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 510ms peak_memory: 1MiB    
Dec 20 11:01:58.143  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 514ms peak_memory: 3MiB    
Dec 20 11:02:07.887  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 539ms peak_memory: 4MiB    
Dec 20 11:02:10.534  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 550ms peak_memory: 2MiB    
Dec 20 11:02:11.026  INFO near_performance_metrics::actix: Slow function call run_later:2 chain/network/src/peer_manager.rs:580 took: 3304ms    
Dec 20 11:02:12.052  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 500ms peak_memory: 1MiB    
Dec 20 11:02:12.588  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 507ms peak_memory: 3MiB    
Dec 20 11:02:13.357  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 718ms peak_memory: 2MiB    
Dec 20 11:02:14.287  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 543ms peak_memory: 3MiB    
Dec 20 11:02:14.924  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 636ms peak_memory: 2MiB    
Dec 20 11:02:15.578  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 637ms peak_memory: 3MiB    
Dec 20 11:02:16.730  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 500ms peak_memory: 1MiB    
Dec 20 11:02:18.191  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 508ms peak_memory: 3MiB    
Dec 20 11:02:18.821  INFO near_performance_metrics::stats: Function exceeded time limit near_client::client_actor::ClientActor:1 "near_network::types::NetworkClientMessages" took: 579ms peak_memory: 3MiB    
Dec 20 11:02:19.239  INFO near_performance_metrics::stats: Performance stats 38 threads    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:     Thread:1 ratio: 0.9088 {"near_client::sync::StateSync", "near_client::client_actor::ClientActor", "run_later", "near_telemetry::TelemetryActor", "telemetry"}:0     
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:737 cnt: 30 total: 11ms max: 0ms    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:1182 cnt: 25 total: 27ms max: 18ms    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:1383 cnt: 123 total: 1681ms max: 534ms    
Dec 20 11:02:19.240  INFO near_performance_metrics::stats:         func 0 chain/client/src/client_actor.rs:1396 cnt: 2 total: 40ms max: 30ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 chain/client/src/sync.rs:895 cnt: 3593 total: 509ms max: 34ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 chain/telemetry/src/lib.rs:70 cnt: 14 total: 15ms max: 9ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 near_network::types::NetworkClientMessages:0 cnt: 3460 total: 52242ms max: 1166ms    
Dec 20 11:02:19.241  INFO near_performance_metrics::stats:         func 0 near_telemetry::TelemetryEvent:0 cnt: 2 total: 1ms max: 0ms    
....
```

**Comment by bowenwang1996** [comment]
Discussed offline. This happens when the node is syncing and is expected. See #3262

**Comment by pmnoxx** [comment]
I've been running a node for last 2 weeks, and I haven't seen any other memory spikes since that time.

**Comment by bowenwang1996** [comment]
@pmnoxx I did a bit more investigation with the help of `--memory_stats` and found that it seems that this is indeed caused by sending state parts. For some reason, the memory allocated in peer actor is not deallocated:

```
[2mFeb 17 19:20:26.893[0m [33m WARN[0m near_performance_metrics::stats_enabled:     Thread:4484 ratio: 0.003 {"run_later", "near_network::peer::Peer"}:16932 memory: 214MiB(66)
[2mFeb 17 19:21:26.894[0m [33m WARN[0m near_performance_metrics::stats_enabled:     Thread:4484 ratio: 0.008 {"run_later", "near_network::peer::Peer"}:16932 memory: 425MiB(65) 
[2mFeb 17 19:25:26.898[0m [33m WARN[0m near_performance_metrics::stats_enabled:     Thread:4484 ratio: 0.011 {"run_later", "near_network::peer::Peer"}:16932 memory: 846MiB(65)
[2mFeb 17 19:32:26.904[0m [33m WARN[0m near_performance_metrics::stats_enabled:     Thread:17069 ratio: 0.027 {"near_network::peer_manager::PeerManagerActor", "run_later"}:16932 memory: 46MiB(16651)
[2mFeb 17 19:52:26.920[0m [33m WARN[0m near_performance_metrics::stats_enabled:     Thread:4484 ratio: 0.042 {"run_later", "near_network::peer::Peer"}:16932 memory: 3373MiB(250)
```
Looks like a lot of memory is allocated within 30ms.

**Comment by pmnoxx** [comment]
@bowenwang1996 I see that the memory usage grew on  thread 4484 by 2.3gb over period of 30 minutes.
I'll improve the tool that I use for analyzing memory usage, to have an option to specify an individual thread. That would allow us to see how the memory issued by this Peer.

**Comment by pmnoxx** [comment]
@bowenwang1996 In order to get the tool to work, we would have to run nodes with debug symbols, that shouldn't make instances slower, but will require to change how we deploy nodes.

**Comment by bowenwang1996** [comment]
> I see that the memory usage grew on thread 4484 by 2.3gb over period of 30 minutes, not ms.

Yeah sorry for the mistake. Updated

**Comment by bowenwang1996** [comment]
> In order to get the tool to work, we would have to run nodes with debug symbols, that shouldn't make instances slower, but will require to change how we deploy nodes.

It's probably better if you first try it on one node and if it works well we can deploy this to every node owned by us.

**Comment by pmnoxx** [comment]
Yes, ofcourse.

**Comment by pmnoxx** [comment]
> > In order to get the tool to work, we would have to run nodes with debug symbols, that shouldn't make instances slower, but will require to change how we deploy nodes.
> 
> It's probably better if you first try it on one node and if it works well we can deploy this to every node owned by us.

I added an option to provide memory statistics for a single threads: https://github.com/near/near-memory-tracker/pull/13

**Comment by pmnoxx** [comment]
I started two nodes, we will have to wait until something we see something. I have tools ready, they should show us memory usage per thread.

**Comment by bowenwang1996** [comment]
> I started two nodes

@pmnoxx on which network? It is probably easier to do it on testnet since the state size is larger there

**Comment by pmnoxx** [comment]
@bowenwang1996 I started them on mainnet. Did the  issue you describe happen on mainnet or testnet?

**Comment by bowenwang1996** [comment]
> Did the issue you describe happen on mainnet or testnet?

On testnet

**Comment by pmnoxx** [comment]
@bowenwang1996 

I've seen a case where a node would die, because a peer started growing in memory. 
While calling method SendMessage, it grew by 4gb, then it grew by another 4gb.
Afterwards the whole process died, and I couldn't do a memory dump, because instance has a memory limit of 8gb.

**Comment by pmnoxx** [comment]
@bowenwang1996 I saw a case where a `PeerActor` caused `OOM`, we can see that memory usage grew while allocating `BytesMut` of bigger size.

```

[2mFeb 24 22:07:49.814[0m [33m WARN[0m near_rust_allocator_proxy::allocator: Thread 28285 reached new record of memory usage 71703MiB
   0: <near_rust_allocator_proxy::allocator::MyAllocator as core::alloc::global::GlobalAlloc>::alloc
             at /home/pmnoxx/.cargo/registry/src/github.com-1ecc6299db9ec823/near-rust-allocator-proxy-0.2.9/src/allocator.rs:202:26
   1: __rg_alloc
             at /home/pmnoxx/repos/nearcore.master/neard/src/main.rs:26:1
   2: alloc::alloc::alloc
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:76:14
      alloc::alloc::Global::alloc_impl
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:155:73
   3: <alloc::alloc::Global as core::alloc::AllocRef>::alloc
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:214:9
      alloc::raw_vec::RawVec<T,A>::allocate_in
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:188:45
   4: alloc::raw_vec::RawVec<T,A>::with_capacity_in
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:163:9
      alloc::raw_vec::RawVec<T>::with_capacity
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:93:9
      alloc::vec::Vec<T>::with_capacity
             at /home/pmnoxx/.rustup/toolchains/nightly-2020-10-08-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:363:20
   5: bytes::bytes_mut::BytesMut::reserve_inner
             at /home/pmnoxx/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes_mut.rs:649:39
   6: bytes::bytes_mut::BytesMut::reserve
             at /home/pmnoxx/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes_mut.rs:548:9
   7: <near_network::codec::Codec as tokio_util::codec::encoder::Encoder<alloc::vec::Vec<u8>>>::encode
             at /home/pmnoxx/repos/nearcore.master/chain/network/src/codec.rs:30:13
   8: actix::io::FramedWrite<I,T,U>::write
             at /home/pmnoxx/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-0.11.0-beta.1/src/io.rs:395:17
   9: near_network::peer::Peer::send_message
             at /home/pmnoxx/repos/nearcore.master/chain/network/src/peer.rs:265:17
  10: <near_network::peer::Peer as actix::handler::Handler<near_network::types::SendMessage>>::handle::{{closure}}
             at /home/pmnoxx/repos/nearcore.master/chain/network/src/peer.rs:1032:9
```

**Comment by pmnoxx** [comment]
I also added a print, we see that we tried to send 11000 messages to Peer, which are of type "StateResponse"
```

[2mFeb 24 22:05:19.662[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1891775
[2mFeb 24 22:05:19.664[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1877985
[2mFeb 24 22:05:19.678[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1877937
[2mFeb 24 22:05:19.692[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 3749791
[2mFeb 24 22:05:19.705[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1877231
[2mFeb 24 22:05:19.715[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1877524
[2mFeb 24 22:05:19.726[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 2041138
[2mFeb 24 22:05:19.728[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1877431
[2mFeb 24 22:05:19.743[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 2041452
[2mFeb 24 22:05:19.746[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1863067
[2mFeb 24 22:05:19.748[0m [32m INFO[0m near_network::peer: send message type: StateResponse length: 1814682
```

**Comment by bowenwang1996** [comment]
Should be fixed after #4012 and #4016
